src/geometry_aux.cpp

```diff
@@ -40,6 +42,71 @@ void update_universe_cell_count(int32_t a, int32_t b)
   }
 }
 
+double calculate_optical_thickness(Position start_sampled_position,
+  Position end_sampled_position, uint64_t* seed)
+{
+  Position direction = (end_sampled_position - start_sampled_position) /
+                       (end_sampled_position - start_sampled_position).norm();
+
+  Particle p;
+  SourceSite site;
+  site.E = 1.0;
+  site.particle = ParticleType::neutron;
+  site.r = start_sampled_position;
+  site.u = direction;
+  p.from_source(&site);
+
+  double optical_thickness = 0.0;
+
+  while (true) {
+    if (!exhaustive_find_cell(p))
+      break;
+
+    BoundaryInfo boundary = distance_to_boundary(p);
+    double distance_to_boundary = boundary.distance;
+    double distance_to_endpoint = (end_sampled_position - p.r()).norm();
+
+    // Check if we are in the cell containing the endpoint
+    if (distance_to_endpoint < distance_to_boundary) {
```

Comment:
This stopping condition is a good start.  We'll have to think through some of the pathologies that may occur and write tests to confirm that it works over all those edge cases.